torch/csrc/jit/passes/bailout_graph.cpp

```diff
@@ -0,0 +1,216 @@
+  }
+
+  Value* mapExistingInputForValue(Value* old_value, Value* new_value) {
+    this->old_to_new_[old_value] = new_value;
+    new_value->copyMetadata(old_value);
+    return new_value;
+  }
+
+  Value* getOrAddInputForValue(Value* v) {
+    if (this->old_to_new_.count(v) == 0) {
+      return addNewInputForValue(v);
+    } else {
+      return this->old_to_new_[v];
+    }
+  }
+
+  // buildBailOutBlockFrom builds a bailout graph from
+  // a given node `n` until the end of the owning block
+  // If `n` belongs to `prim::If` or `prim::Loop`
+  // buildBailOutGraphForOwningNode continues
+  // from block's owning node (e.g. `prim::If` or
+  // `prim::Loop`)
+  void buildBailOutBlockFrom(Node* n) {
+    auto* block = copy_graph_->block();
+    auto b = n->owningBlock();
+    graph_node_list_iterator it(n, kNextDirection);
+    for (; it != b->nodes().end(); it++) {
+      auto env = [this](Value* v) { return getOrAddInputForValue(v); };
+
+      auto node = *it;
+
+      auto new_node = block->appendNode(copy_graph_->createClone(node, env));
+      for (size_t i = 0; i < node->outputs().size(); ++i) {
+        auto oo = node->outputs()[i];
+        auto no = new_node->outputs()[i];
+        old_to_new_[oo] = no;
+        no->copyMetadata(oo);
+      }
+    }
+
+    // we are either in `prim::If` or `prim::Loop`
+    // bailout graph building will continue from `outer_node` next
+    // remember/map the outputs needed to unwind this `prim::If`
+    // or `prim::Loop`
+    auto outer_node = n->owningBlock()->owningNode();
+    if (outer_node) {
+      buildBailOutGraphForOwningNode(n, outer_node);
+    }
+  }
+
+  // buildBailOutGraphForOwningNode builds a bailout graph
+  // for a CFG given node () as follows:
+  // * if `inner_node` is in a true or false arm of an `prim::If` node
+  // the rest of the arm gets inlined in the bailout graph
+  // and we continue building the rest of the bailout graph
+  // from the node *following* `outer_node`.
+  // * if `inner_node` is in a loop, the rest of the loop gets inlined
+  // we remember the loop outputs needed to execute the
+  // we adjust the upper limit (old_upper_limit - current_iteration)
+  // note the stop condition is already mapped (remembered)
+  // then, we continue building the rest of the bailout graph
+  // starting from the owning `outer_node`
+  // which implies `outer_node` will be cloned
+  void buildBailOutGraphForOwningNode(Node* inner_node, Node* outer_node) {
+    auto* block = copy_graph_->block();
+    auto block_outputs = inner_node->owningBlock()->outputs();
+    // skip the first input for loops (current iteration count)
+    size_t i = outer_node->kind() == prim::Loop;
+    auto new_outputs = outer_node->kind() == prim::Loop ? outer_node->inputs()
+                                                        : outer_node->outputs();
+    for (; i < block_outputs.size(); i++) {
+      auto nv = old_to_new_[block_outputs[i]];
+      old_to_new_[new_outputs.at(i)] = nv;
```

Comment:
This loop is confusing here because it is actually doing something different for If/Loop. I also think it is mapping the loop condition node to the trip count node I think it would be more clear folded into those cases. This would make it possible to be more careful in copying over the output of the loop block to the input of the loop instruction.